### PR TITLE
Added Axis.NaturalSize (issue #383)

### PR DESCRIPTION
**Pull request by tr8dr**
"Simple" change to expose Axis "natural size" so that application can make adjustments to align panes.

### CODE REVIEW

**Comment by objorke** [comment]
#383 (link to issue should be in PR description, not title)

**Comment by objorke** [comment]
It should not be 197 commits in this PR. Please rebase to latest before making a PR

File: Source/OxyPlot/Axes/Axis.cs [hunk 1]
```diff
@@ -543,6 +543,11 @@ public double Scale
         public bool UseSuperExponentialFormat { get; set; }
 
         /// <summary>
+        /// Gets or sets the natural dimensions of the axis, as of most recent rendering.
+        /// </summary>
+        public OxySize NaturalSize { get; protected set; }
```
Comment (objorke) [inline Source/OxyPlot/Axes/Axis.cs]: I suggest "ActualSize" for consistency

File: Source/OxyPlot/Axes/Axis.cs [hunk 2]
```diff
@@ -835,85 +840,13 @@ public bool IsVertical()
...
 
         /// <summary>
-        /// Measures the size of the axis (maximum axis label width/height).
+        /// Updates the natural dimensions of axis with respect to graphics context and
+        /// any other graphical housekeeping.
         /// </summary>
         /// <param name="rc">The render context.</param>
-        /// <returns>The size of the axis.</returns>
-        public virtual OxySize Measure(IRenderContext rc)
+        public void Update(IRenderContext rc)
```
Comment (objorke) [inline Source/OxyPlot/Axes/Axis.cs]: do not expose this - find other solution

File: Source/OxyPlot/Axes/Axis.cs [hunk 3]
```diff
@@ -1353,6 +1286,88 @@ protected internal virtual void ResetCurrentValues()
         }
 
         /// <summary>
+        /// Measures the size of the axis (maximum axis label width/height).
+        /// </summary>
+        /// <param name="rc">The render context.</param>
+        /// <returns>The size of the axis.</returns>
+        protected virtual OxySize Measure(IRenderContext rc)
```
Comment (objorke) [inline Source/OxyPlot/Axes/Axis.cs]: yes, I think it is a good idea to change to protected

File: Source/OxyPlot/PlotModel/PlotModel.Rendering.cs [hunk 4]
```diff
@@ -179,7 +179,10 @@ private static double MaxSizeOfPositionTier(IRenderContext rc, IEnumerable<Axis>
             double maxSizeOfPositionTier = 0;
             foreach (var axis in axesOfPositionTier)
             {
-                var size = axis.Measure(rc);
+                // update axis with latest render context, adjusting dimensions, etc.
+                axis.Update(rc);
```
Comment (objorke) [inline Source/OxyPlot/PlotModel/PlotModel.Rendering.cs]: can we find some other solution?

File: Source/OxyPlot/Axes/Axis.cs (hunk 1 above)
Comment (tr8dr) [inline Source/OxyPlot/Axes/Axis.cs]: Isn't ActualSize used to indicate the size that was chosen for rendering?  NaturalSize indicates the "natural" size of the element if involved no clipping / size override.   Correct me if I am wrong, if size is overriden by the user then Actual variables will reflect the override.

File: Source/OxyPlot/Axes/Axis.cs (hunk 2 above)
Comment (tr8dr) [inline Source/OxyPlot/Axes/Axis.cs]: protected internal?   (don't know if you need internal or not)?

File: Source/OxyPlot/PlotModel/PlotModel.Rendering.cs (hunk 4 above)
Comment (tr8dr) [inline Source/OxyPlot/PlotModel/PlotModel.Rendering.cs]: What would you suggest?  I just need a property on Axes which reflects the "desired" or "natural" size (as opposed to the rendered size which may reflect user size overrides).

File: Source/OxyPlot/Axes/Axis.cs (hunk 1 above)
Comment (objorke) [inline Source/OxyPlot/Axes/Axis.cs]: Right, the "ActualX" properties should indicate the actual size that was chosen for rendering.
Sorry, I don't get the purpose of this new property yet...

File: Source/OxyPlot/PlotModel/PlotModel.Rendering.cs (hunk 4 above)
Comment (objorke) [inline Source/OxyPlot/PlotModel/PlotModel.Rendering.cs]: or minimum/maximum size?

**Comment by objorke** [comment]
Need an example that shows the necessity of the new property.

**Comment by tr8dr** [comment]
I put an example in the issue, where explained why I need it.  Do you need more clarification?

**Comment by objorke** [comment]
The example should be runnable and included in the PR

**Comment by tr8dr** [comment]
Given that aligning 2 different panes is specific to a particular rendering library (Gtk, WPF, etc) I cannot provide a visual example for this.  The important thing is that the property exists and has proper values.  The best I can do is have the property value render in the legend or something like that?  What would you like to see?

File: Source/OxyPlot/Axes/Axis.cs (hunk 1 above)
Comment (tr8dr) [inline Source/OxyPlot/Axes/Axis.cs]: re: rebase, i tried that.  When I submitted this saw that included intermediate commits (sorry).  I do not know why the intermediate changes in the main repo show up here.  Obviously I did something wrong.  I followed the following guide:  

http://robots.thoughtbot.com/keeping-a-github-fork-updated

**Comment by objorke** [comment]
If we have no example where we can test that this actually works, this feature may easily be broken in the future. A unit test that asserts on values is ok, I think that can be done in portable code.
An platform specific example in Gtk or any of the other platforms would also be great to debug and review the solution!

**Comment by tr8dr** [comment]
I can make the property visible as text in a plot.  That would work in the ExamplesLibrary without issue (indeed this is how I tested it initially).

**Comment by objorke** [comment]
Text in the plot is not very helpful, it will be very difficult to see regressions on the number in the future.

Sorry for being a bit difficult on this issue, but I am concerned about the maintainability of the library. The layout engine is one part I would really like to see being refactored and improved.

**Comment by tr8dr** [comment]
I agree with the point.  How do you test rendering, which involves a graphics context and event loop though?  Do you have this rigged for testing?   I took a look at OxyPlot.Tests but did not notice that sort of unit test.

**Comment by tr8dr** [comment]
If you can point me to a unit test example that would work here, greatly appreciated.   I like to unit test my libraries as well.  I have never done one for UIs though.

File: Source/OxyPlot/Axes/Axis.cs (hunk 2 above)
Comment (objorke) [inline Source/OxyPlot/Axes/Axis.cs]: I think the `DesiredSize` property can be set during the call to `Measure` - I understand this is how it is done in WPF. Revert (remove) the Update method.

**Comment by objorke** [comment]
Rendering a `PlotModel` on a Svg or Pdf context should be possible in the OxyPlot.Tests project. After the `PlotModel` has been updated/rendered, it should be possible to assert the `Axis.DesiredSize` property.

Personally, I would have created a example in the [WpfExamples](https://github.com/oxyplot/oxyplot/tree/develop/Source/Examples/WPF/WpfExamples/Examples) project showing exactly how you would use the new property in your application. This is probably not much more work than writing a unit test, and can be a good demo on how to use the property. It also means that the use case you suggest is more likely to be supported in the future! :-)

**Comment by tr8dr** [comment]
This branch is a mess because of my failed rebase.  Let me close this pull-request and see whether I can make proper and re-issue.   I am still not sure how to rebase the fork properly.  What I did last time apparently did not work ...

**Comment by objorke** [comment]
See #386 for unit tests and wpf example

**Comment by tr8dr** [comment]
I can't do a WPF (since I run on OS X), but I should be able to do something with pdf ...

**Pull request closed without merging**